pkg/objectstore/migrate.go

```diff
@@ -0,0 +1,330 @@
+package objectstore
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/minio/minio-go/v7"
+	"github.com/minio/minio-go/v7/pkg/credentials"
+	"github.com/replicatedhq/ekco/pkg/cluster/types"
+	"github.com/replicatedhq/ekco/pkg/util"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// SyncAllBuckets syncs copies all objects in all buckets in object store to another, and returns progress via a channel.
+func SyncAllBuckets(ctx context.Context, sourceEndpoint, sourceAccessKey, sourceSecretKey, destEndpoint, destAccessKey, destSecretKey string, logs chan<- string) error {
+	// Initialize source client object.
+	minioClient, err := minio.New(sourceEndpoint, &minio.Options{
+		Creds: credentials.NewStaticV4(sourceAccessKey, sourceSecretKey, ""),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to initialize minio client: %v", err)
+	}
+
+	// Initialize dest client object.
+	rookClient, err := minio.New(destEndpoint, &minio.Options{
+		Creds: credentials.NewStaticV4(destAccessKey, destSecretKey, ""),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to initialize rook client: %v", err)
+	}
+
+	if logs != nil {
+		logs <- "Initialized clients"
+	}
+
+	minioBuckets, err := minioClient.ListBuckets(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to list minio buckets: %v", err)
+	}
+
+	if logs != nil {
+		logs <- fmt.Sprintf("Found %d buckets to sync", len(minioBuckets))
+	}
+
+	for _, bucket := range minioBuckets {
+		if logs != nil {
+			logs <- fmt.Sprintf("Syncing objects in %s", bucket.Name)
+		}
+
+		startTime := time.Now()
+		numObjects, err := syncBucket(ctx, minioClient, rookClient, bucket.Name, logs)
+		if err != nil {
+			return fmt.Errorf("failed to sync bucket %s: %v", bucket.Name, err)
+		}
+
+		if logs != nil {
+			logs <- fmt.Sprintf("Copied %d objects in %s over %s", numObjects, bucket.Name, time.Since(startTime).String())
+		}
+	}
+
+	return nil
+}
+
+// UpdateConsumers updates the access key and secret key for all consumers of the object store.
+// it handles kotsadm, registry, and velero.
+func UpdateConsumers(ctx context.Context, controllers types.ControllerConfig, endpoint, hostname, accessKey, secretKey, originalHostname, originalSecretKey string, logs chan<- string) error {
+	client := controllers.Client
+	// if kubernetes_resource_exists default secret kotsadm-s3
+	err := updateKotsadmObjectStore(ctx, client, logs, accessKey, secretKey, hostname, endpoint)
+	if err != nil {
+		return err
+	}
+
+	err = updateRegistryObjectStore(ctx, client, logs, accessKey, secretKey, hostname, endpoint)
+	if err != nil {
+		return err
+	}
+
+	err = updateVeleroObjectStore(ctx, controllers, logs, accessKey, secretKey, hostname, endpoint, originalSecretKey, originalHostname)
```

Comment:
I don't think it would roll back kots data, as kots just adds new files and doesn't update them
but rearranging these is easy

Reply:
i dont think my concern is warranted here. one thing to note though is during the data migration couldnt these consumers continue to write as we dont disable them?

Reply:
we do disable them - the first thing we do before migrating data is break the minio service selector (by adding 'doesnotexist') so that it doesn't point at anything